consensus/application.go

```diff
@@ -339,124 +341,156 @@ func dupProof(se *types.StateElement) {
+	} else if fced.Revision != nil {
+		*fced.Revision = rev
 	} else {
-		if r, ok := ms.revs[fce.ID]; ok {
-			r.FileContract = rev
-		} else {
-			// store the original
-			dupProof(&fce.StateElement)
-			ms.fces = append(ms.fces, fce)
-			// store the revision
-			dupProof(&fce.StateElement)
-			fce.FileContract = rev
-			ms.revs[fce.ID] = &fce
-		}
+		fced.FileContractElement = fce
+		fced.Revision = &rev
 	}
 }
 
 func (ms *MidState) resolveFileContractElement(fce types.FileContractElement, valid bool, txid types.TransactionID) {
-	ms.res[fce.ID] = valid
-	ms.spends[fce.ID] = txid
+	fced := ms.recordFileContractElement(fce.ID)
+	fced.Resolved = true
+	fced.Valid = valid
 	dupProof(&fce.StateElement)
-	ms.fces = append(ms.fces, fce)
+	fced.FileContractElement = fce
+	ms.spends[fce.ID] = txid
 }
 
-func (ms *MidState) addV2FileContractElement(id types.FileContractID, fc types.V2FileContract) {
-	fce := types.V2FileContractElement{
+func (ms *MidState) recordV2FileContractElement(id types.FileContractID) *V2FileContractElementDiff {
+	if i, ok := ms.elements[id]; ok {
+		return &ms.v2fces[i]
+	}
+	ms.v2fces = append(ms.v2fces, V2FileContractElementDiff{})
+	ms.elements[id] = len(ms.v2fces) - 1
+	return &ms.v2fces[len(ms.v2fces)-1]
+}
+
+func (ms *MidState) createV2FileContractElement(id types.FileContractID, fc types.V2FileContract) {
+	fced := ms.recordV2FileContractElement(id)
+	fced.V2FileContractElement = types.V2FileContractElement{
 		StateElement:   types.StateElement{LeafIndex: types.UnassignedLeafIndex},
 		ID:             id,
 		V2FileContract: fc,
 	}
-	ms.v2fces = append(ms.v2fces, fce)
-	ms.created[ms.v2fces[len(ms.v2fces)-1].ID] = len(ms.v2fces) - 1
-	ms.siafundTaxRevenue = ms.siafundTaxRevenue.Add(ms.base.V2FileContractTax(fce.V2FileContract))
+	fced.Created = true
+	ms.siafundTaxRevenue = ms.siafundTaxRevenue.Add(ms.base.V2FileContractTax(fc))
 }
 
 func (ms *MidState) reviseV2FileContractElement(fce types.V2FileContractElement, rev types.V2FileContract) {
-	if i, ok := ms.created[fce.ID]; ok {
-		ms.v2fces[i].V2FileContract = rev
+	fced := ms.recordV2FileContractElement(fce.ID)
+	if fced.Created {
+		fced.V2FileContractElement.V2FileContract = rev
+	} else if fced.Revision != nil {
+		*fced.Revision = rev
 	} else {
-		if r, ok := ms.v2revs[fce.ID]; ok {
-			r.V2FileContract = rev
-		} else {
-			// store the original
-			dupProof(&fce.StateElement)
-			ms.v2fces = append(ms.v2fces, fce)
-			// store the revision
-			dupProof(&fce.StateElement)
-			fce.V2FileContract = rev
-			ms.v2revs[fce.ID] = &fce
-		}
+		fced.V2FileContractElement = fce
+		fced.Revision = &rev
 	}
 }
 
 func (ms *MidState) resolveV2FileContractElement(fce types.V2FileContractElement, res types.V2FileContractResolutionType, txid types.TransactionID) {
-	ms.v2res[fce.ID] = res
-	ms.spends[fce.ID] = txid
+	fced := ms.recordV2FileContractElement(fce.ID)
+	if fced.Created {
+		panic("consensus: resolved a newly-created v2 contract")
```

Comment:
Do we ever update the `created` bool of a fce in the midstate? I'm sure it's fine but since it's a newly introduced check (and panic) I wanted to draw attention to it to be 100% sure it's valid.

Reply:
no; `Created` is a field of t`FileContractElementDiff`, not `FileContractElement`, and once it's set to `true` we never touch it again.